### PR TITLE
feat: only interpolate system parameter in component configuration section, not configuration parameter

### PR DESCRIPTION
**Pull request by rightpeter**
**Description of changes:**
Before, nucleus interpolates component config after aggregating config from recipe, deployment update and running config. Running config is dynamic which exposes the risk of attacker can manipulate the running config in a way to be interpolated in next deployment.

This change let nucleus only interpolate system parameter for component configuration, not configuration parameter. For example:
* `{configuration:/paramA}` won't be interpolated in configuration section
* `{iot:thingName}` will be interpolated
This makes the component configuration section interpolation behavior static.

**Why is this change necessary:**
Security risk: enabling interpolation for configuration parameter under configuration section can cause issue like
corrupted running config triggers unexpected behavior by having special combination of variable designator.

**How was this change tested:**
- [X] Updated or added new unit tests.
- [ ] Updated or added new integration tests.
- [ ] Updated or added new end-to-end tests.
- [ ] If my code makes a remote network call, it was tested with a proxy.

**Any additional information or context required to review the change:**

**Documentation Checklist:**
 - [ ] Updated the README if applicable.

**Compatibility Checklist:**
- [X] I confirm that the change is backwards compatible.
- [ ] Any modification or deletion of public interfaces does not impact other plugin components.
- [X] For external library version updates, I have reviewed its change logs and Nucleus does not consume 
  any deprecated method or type.

Refer to [Compatibility Guidelines](/COMPATIBILITY.md) for more information.

By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of your choice.

### CODE REVIEW

File: src/main/java/com/aws/greengrass/componentmanager/KernelConfigResolver.java [hunk 1]
```diff
@@ -505,22 +543,22 @@ private String replace(String stringValue, ComponentIdentifier componentIdentifi
...
-            if (CONFIGURATION_NAMESPACE.equals(namespace)) {
+            if (shouldInterpolateConfiguration && CONFIGURATION_NAMESPACE.equals(namespace)) {
                 Optional<String> configReplacement =
                         lookupConfigurationValueForComponent(targetComponent, key, resolvedKernelServiceConfig);
                 if (configReplacement.isPresent()) {
                     stringValue = stringValue.replace(matcher.group(), configReplacement.get());
                 }
-
-            } else if (systemParameters.containsKey(namespace)) {
+            } else if (!Objects.isNull(resolvedKernelServiceConfig) && systemParameters.containsKey(namespace)) {
```
Comment (MikeDombo) [inline src/main/java/com/aws/greengrass/componentmanager/KernelConfigResolver.java:561]: Why is this needed now?

File: src/test/java/com/aws/greengrass/componentmanager/KernelConfigResolverTest.java [hunk 2]
```diff
@@ -481,66 +481,6 @@ void GIVEN_deployment_with_artifact_and_interpolate_config_set_WHEN_config_resol
                 equalTo("java -jar " + "{artifacts:{randomNameSpace:" + artifactsPath + "}}" + "/test.jar -x arg"));
     }
 
-    @Test
```
Comment (MikeDombo) [inline src/test/java/com/aws/greengrass/componentmanager/KernelConfigResolverTest.java:484]: Keep this test, just disable it with a junit annotation

File: src/main/java/com/aws/greengrass/componentmanager/KernelConfigResolver.java [hunk 3]
```diff
@@ -555,8 +593,8 @@ private Optional<String> lookupConfigurationValueForComponent(String componentNa
 
         Map componentResolvedConfig;
 
-        if (resolvedKernelServiceConfig.containsKey(componentName) && ((Map) resolvedKernelServiceConfig
-                .get(componentName)).containsKey(CONFIGURATION_CONFIG_KEY)) {
+        if (!Objects.isNull(resolvedKernelServiceConfig) && resolvedKernelServiceConfig.containsKey(componentName)
```
Comment (MikeDombo) [inline src/main/java/com/aws/greengrass/componentmanager/KernelConfigResolver.java:596]: Why is this now required?

File: src/main/java/com/aws/greengrass/componentmanager/KernelConfigResolver.java [hunk 4]
```diff
@@ -416,6 +417,22 @@ private List<String> pointerToPath(JsonPointer pointer) {
...
+     * @param componentIdentifier         target component id
+     * @param dependencies                name set of component's dependencies
+     * @param resolvedKernelServiceConfig resolved kernel configuration under "Services" key
+     * @return the interpolated lifecycle object
+     * @throws IOException for directory issues
+     */
+    public Object interpolateSystemParametersOnly(Object configValue, ComponentIdentifier componentIdentifier,
+                                                 Set<String> dependencies,
+                                                  Map<String, Object> resolvedKernelServiceConfig) throws IOException {
+        return interpolate(configValue, componentIdentifier, dependencies, resolvedKernelServiceConfig, false);
```
Comment (MikeDombo) [inline src/main/java/com/aws/greengrass/componentmanager/KernelConfigResolver.java:433]: I don't really like the idea of adding this boolean. Perhaps pass in the map of interpolatable things instead, and in this case copy the map and remove configuration.

File: src/main/java/com/aws/greengrass/componentmanager/KernelConfigResolver.java (hunk 1 above)
Comment (rightpeter) [reply src/main/java/com/aws/greengrass/componentmanager/KernelConfigResolver.java:561]: Just do the validation before calling. Is there any concern to do so?

Comment (MikeDombo) [reply src/main/java/com/aws/greengrass/componentmanager/KernelConfigResolver.java:561]: Well it is a new check which was not considered necessary before. If the value is null it seems that we have some massive problems. I think we would want it to fail here.

Comment (MikeDombo) [reply src/main/java/com/aws/greengrass/componentmanager/KernelConfigResolver.java:561]: Can the value ever be null?

File: src/test/java/com/aws/greengrass/componentmanager/KernelConfigResolverTest.java (hunk 2 above)
Comment (rightpeter) [reply src/test/java/com/aws/greengrass/componentmanager/KernelConfigResolverTest.java:484]: Changed it to Disabled.

File: src/main/java/com/aws/greengrass/componentmanager/KernelConfigResolver.java (hunk 4 above)
Comment (rightpeter) [reply src/main/java/com/aws/greengrass/componentmanager/KernelConfigResolver.java:433]: This boolean make it explicit that if it's interpolate all parameters or only the system parameters which is more deterministic and readable. If we copy the map and remove configuration, it's more dynamic and expose the interpolation behavior to the caller. What's the benefit of doing so?

File: src/main/java/com/aws/greengrass/componentmanager/KernelConfigResolver.java (hunk 1 above)
Comment (rightpeter) [reply src/main/java/com/aws/greengrass/componentmanager/KernelConfigResolver.java:561]: With our current code, not likely. The `resolvedKernelServiceConfig` was initialized at the very beginning.

But `resolve` is a public method, in the future it's possible that someone pass an empty one.

Comment (MikeDombo) [reply src/main/java/com/aws/greengrass/componentmanager/KernelConfigResolver.java:561]: If prefer that you undo this.

File: src/main/java/com/aws/greengrass/componentmanager/KernelConfigResolver.java (hunk 4 above)
Comment (MikeDombo) [reply src/main/java/com/aws/greengrass/componentmanager/KernelConfigResolver.java:433]: A free floating boolean is more difficult to understand than using an enum or even what I'm suggesting with the map. I expect that this change will be temporary, so I'm OK with this for now. If it stays longer, I'd like a more explicit option.

**Review by shaguptashaikh — APPROVED** [review]

**Review by MikeDombo — APPROVED** [review]
